racket/collects/racket/private/dict.rkt

```diff
@@ -244,23 +196,39 @@
       (raise-argument-error 'dict-map "dict?" d))
     (proc (car x) (cdr x))))
 
+(define (assoc-fold-unique f init d)
+  (unless (assoc? d)
+    (raise-argument-error 'assoc-fold-unique "dict?" d))
+  (let loop ([xd d]
+             [acc init]
+             [seen (make-immutable-hash)])
```

Comment:
It might be nice to implement the optimization from `remove-duplicates` to use a plain list for seen when the given list is less than length 40 (with `member` and `cons` instead of `hash-has-key?` and `hash-set`), and perhaps also to special-case lists of length ≤1.

Reply:
Can you provide benchmarks (or larger programs) to check whether this would pay off?

(I'm not even sure that the previous change, removing `remove-duplicates`, is going to save much time on the programs that end up using alist dicts.)

Reply:
As I think about @samth's comment, one way of considering it would be that, if you are using alist dicts, you've made a choice for some reason to not use hash tables, and you've done that knowing the performance trade-offs (e.g. potentially faster for small dicts, likely much slower for very big ones). That's pushing me further toward thinking this code should use `member` and `cons` rather than hash-tables (as `remove-duplicates` does for small-ish lists, and as @Bogdanp also suggested). I might even skip the `(if (< (length d) 40) ... ...)` check, on the grounds that there is an explicit choice here not to use hash-tables.

Reply:
But overall, I think having the dict methods behave correctly is a much bigger deal than too much performance tuning. Especially for alists-as-dicts, you've already signed up to pay for generic dispatch and likely to get less inlining, which many of the list-specific functions go to great lengths to promote.

Reply:
Is it ok to merge this, then, and leave the tuning for another patch?

Reply:
Sounds good to me! (I don't have commit access.)